File: db4o.net/Db4objects.Db4o/Db4objects.Db4o/Internal/References/TransactionalReferenceSystemBase.cs
/* Copyright (C) 2004 - 2011  Versant Inc.  http://www.db4o.com */

using Db4objects.Db4o.Foundation;
using Db4objects.Db4o.Internal;
using Db4objects.Db4o.Internal.References;

namespace Db4objects.Db4o.Internal.References
{
	/// <exclude></exclude>
	public abstract class TransactionalReferenceSystemBase
	{
		protected readonly IReferenceSystem _committedReferences;

		protected IReferenceSystem _newReferences;

		public TransactionalReferenceSystemBase()
		{
			CreateNewReferences();
			_committedReferences = NewReferenceSystem();
		}

		private IReferenceSystem NewReferenceSystem()
		{
			return new HashcodeReferenceSystem();
		}

		public abstract void AddExistingReference(ObjectReference @ref);

		public abstract void AddNewReference(ObjectReference @ref);

		public abstract void Commit();

		protected virtual void TraverseNewReferences(IVisitor4 visitor)
		{
			_newReferences.TraverseReferences(visitor);
		}

		protected virtual void CreateNewReferences()
		{
			_newReferences = NewReferenceSystem();
		}

		public virtual ObjectReference ReferenceForId(int id)
		{
			ObjectReference @ref = _newReferences.ReferenceForId(id);
			if (@ref != null)
			{
				return @ref;
			}
			return _committedReferences.ReferenceForId(id);
		}

		public virtual ObjectReference ReferenceForObject(object obj)
		{
			ObjectReference @ref = _newReferences.ReferenceForObject(obj);
			if (@ref != null)
			{
				return @ref;
			}
			return _committedReferences.ReferenceForObject(obj);
		}

		public abstract void RemoveReference(ObjectReference @ref);

		public abstract void Rollback();

		public virtual void TraverseReferences(IVisitor4 visitor)
		{
			TraverseNewReferences(visitor);
			_committedReferences.TraverseReferences(visitor);
		}
	}
}

File: db4oj/core/src/com/db4o/internal/references/TransactionalReferenceSystemBase.java
/* Copyright (C) 2009  Versant Corp.  http://www.db4o.com */

package com.db4o.internal.references;

import com.db4o.foundation.*;
import com.db4o.internal.*;

/**
 * @exclude
 */
public abstract class TransactionalReferenceSystemBase {
	
	protected final ReferenceSystem _committedReferences;
	
	protected ReferenceSystem _newReferences;
	
	public TransactionalReferenceSystemBase() {
		createNewReferences();
		_committedReferences = newReferenceSystem();
	}
	
	private ReferenceSystem newReferenceSystem(){
	    return new HashcodeReferenceSystem();
	}

	public abstract void addExistingReference(ObjectReference ref);

	public abstract void addNewReference(ObjectReference ref);
	
	public abstract void commit();

	protected void traverseNewReferences(final Visitor4 visitor) {
		_newReferences.traverseReferences(visitor);
	}
	
	protected void createNewReferences(){
		_newReferences = newReferenceSystem();
	}

	public ObjectReference referenceForId(int id) {
		ObjectReference ref = _newReferences.referenceForId(id);
		if(ref != null){
			return ref;
		}
		return _committedReferences.referenceForId(id);
	}

	public ObjectReference referenceForObject(Object obj) {
		ObjectReference ref = _newReferences.referenceForObject(obj);
		if(ref != null){
			return ref;
		}
		return _committedReferences.referenceForObject(obj);
	}

	public abstract void removeReference(ObjectReference ref);
	
	public abstract void rollback();
	
	public void traverseReferences(Visitor4 visitor) {
		traverseNewReferences(visitor);
		_committedReferences.traverseReferences(visitor);
	}



}
